content/300-guides/100-prisma-guides/100-prisma-client-transactions-guide.mdx

```diff
@@ -0,0 +1,778 @@
+model User {
+  id   Int    @id @default(autoincrement())  
+  id   String  @id @default(uuid())
+  email String @unique
+  teams Team[]
+}
+```
+
+Refactor the sign-up flow example to use the `$transaction` API instead of nested writes:
+
+```ts
+import { v4 } from 'uuid'
+
+const teamID = v4();
+const userID = v4();
+
+await prisma.$transaction([
+    prisma.user.create({
+      data: {
+        id: userID,
+        email: 'alice@prisma.io',
+        team: {
+          id: teamID
+        }
+      }
+    }),
+    client.team.create({
+      data: {
+        id: teamID,
+        name: 'Aurora Adventures'
+      }
+    })
+])
+```
+
+Technically you can still use nested writes with pre-computed APIs if you prefer that syntax:
+
+```ts
+import { v4 } from 'uuid'
+
+const teamID = v4();
+const userID = v4();
+
+await client.team.create({
+  data: {
+    id: teamID,    
+    name: 'Aurora Adventures',
+    members: {
+      create: {
+        id: userID,
+        email: 'alice@prisma.io',
+        team: {
+          id: teamID
+        }
+      },
+    },
+  },
+})
+```
+
+There's no compelling reason to switch to manually generated IDs and the `$transaction` API if you are already using auto-generated IDs and nested writes.
+
+## Read, modify, write
+
+In some cases you may need to perform custom logic as part of an atomic operation - also known as the [read-modify-write pattern](https://en.wikipedia.org/wiki/Read%E2%80%93modify%E2%80%93write). The following is an example of the read-modify-write pattern:
+
+* Read a value from the database
+* Run some logic to manipulate that value (for example, contacting an external API)
+* Write the value back to the database
+
+All operations should **succeed or fail together** without making unwanted changes to the database, but you do not necessarily need to use an actual database transaction. This section of the guide describes two ways to work with Prisma Client and the read-modify-write pattern:
+
+* Designing idempotent APIs
+* Optimistic concurrency control
+
+### Idempotent APIs
+
+Idempotency is the ability to run the same logic with the same parameters multiple times with the same result: the **effect on the database** is the same whether you run the logic one or one thousand times. For example:
+
+* **NOT IDEMPOTENT**: Create a user in the database with email address "letoya@prisma.io". The `User` table **does not** enforce unique email addresses. The effect on the database is different if you run the logic once (one user created) or ten times (ten users created).
+* **IDEMPOTENT**: Create a user in the database with the email address "letoya@prisma.io". The `User` table **does** enforce unique email addresses. The effect on the database is the same if you run the logic once (one user created) or ten times (subsequent attempts to create a user with the same email address will fail due to `UNIQUE` constraint).
+
+Idempotency is something you can actively design, and is not always needed or possible.
```

Comment:
I'd maybe add that if you can pull off idempotency you should try. Operationally it will make your life a lot easier. But sometimes you can't and that's ok.